services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientRolesCondition.java

```diff
@@ -61,13 +60,10 @@ public ClientPolicyVote applyPolicy(ClientPolicyContext context) throws ClientPo
     private boolean isRolesMatched(ClientModel client) {
         if (client == null) return false;
 
-        List<String> rolesForMatching = getRolesForMatching();
+        Stream<String> rolesForMatching = getRolesStreamForMatching();
         if (rolesForMatching == null) return false;
 
-        client.getRoles().stream().forEach(i -> ClientPolicyLogger.log(logger, "client role = " + i.getName()));
-        rolesForMatching.stream().forEach(i -> ClientPolicyLogger.log(logger, "roles expected = " + i));
-
-        boolean isMatched = rolesForMatching.stream().anyMatch(i->client.getRoles().stream().anyMatch(j->j.getName().equals(i)));
+        boolean isMatched = rolesForMatching.anyMatch(i->client.getRolesStream().anyMatch(j->j.getName().equals(i)));
```

Comment:
This line seem to have performance bottleneck `client.getRolesStream()` stream is opened for each element of `rolesForMatching`

I would propose something of (using Sets from Guava) as it may improve readabiliy - `Sets.intersection` instead of `anyMatch(i -> anyMatch)`:
```
private boolean isRolesMatched(ClientModel client) {
        if (client == null) return false;

        Set<String> rolesForMatching = getRolesStreamForMatching();
        if (rolesForMatching == null) return false;

        Set<String> clientRoles = client.getRolesStream().map(RoleModel::getName).collect(Collectors.toSet());
        boolean isMatched = !Sets.intersection(clientRoles, rolesForMatching).isEmpty();
        
        if (isMatched) {
            ClientPolicyLogger.log(logger, "role matched.");
        } else {
            ClientPolicyLogger.log(logger, "role unmatched.");
        }
        return isMatched;
    }

    private Set<String> getRolesStreamForMatching() {
        if (componentModel.getConfig() == null) return null;
        List<String> roles = componentModel.getConfig().get(ClientRolesConditionFactory.ROLES);
        if (roles == null) {
            return null;
        }

        return new HashSet<>(roles);
    }
```

Reply:
It seems that `Sets` is not used in the current keycloak. If `Sets` is introduced, this PR is the first time to use `Sets`.
@mposolda Should we introduce `Sets` to improve the performance of set operation or not?

Reply:
It can be done without `Sets`
```
 private boolean setsIntersect(Set<String> first, Set<String> second) {
        return new HashSet<>(first).removeAll(second);
    }
```

Reply:
@valb3r I see, I'll use `Set.removeAll` for set operation.